### DISCUSSION TITLE
Static (render-once) Livewire components?

### DISCUSSION THREAD

**Discussion by james2doyle** [livewire/livewire#8691] [3. Feature ideas]
TL;DR - let me disable `wire:*` attributes on my components

I'm working on a larger project using the latest Laravel and Filament. Nothing too spectacular, but we do have a lot of Blocks (Filament content modules) that are used all over the content portions of the site.

Right now, I have those blocks mixed amongst partials, Blade components without classes, Blade components with classes, and Livewire components. It's very inconsistent, and there's not really a good way to articulate why they're all different, other than interactive stuff with the server needs Livewire.

I find that the editing experience is much nicer when the Blade components are written with their class counterpart. So I prefer to do that most of the time.

I have so many different "styles" of components spread across so many different folders and files now, though. I find it kind of messy.

The Blade components require you to use `__construct` for props, the Livewire components require you to use `mount`, and there are a few other differences. But the fact that they are different is somewhat of a sticking point for me.

I'd really just prefer to write everything in Livewire because it's consistent, and if you're someone new joining the project, you don't have to worry about understanding the difference between the two. You can learn one and stick with it.

It also prepares us for any interactivity between the server and the client that might appear in new blocks as we refactor or improve them.

However, most of our components are just static HTML output. It's content. Text, images, and then maybe some JavaScript-sprinkled Alpine stuff, like accordions or sliders. They wouldn't use any of the Livewire magic.

So what? Why not just use Livewire and be done with it?

Well, I started to do that, but I noticed that the output of Livewire is not really appropriate for how I'm using it in this case.

Livewire adds all these HTML attributes (`wire:snapshot`, `wire:effects`, `wire:id`) to your parent HTML element, and that ends up making the output of these "static components" quite messy, and sometimes the snapshot will expose information you don't really want it to.

Here is a simple example:

```php
<?php

namespace App\Livewire;

use Livewire\Component;

class StaticElem extends Component
{
    public string $message;

    public function render()
    {
        // no interactivity, just using Livewire for rendering purposes and a consistent API across all components
        return <<<'HTML'
            <div>This is a static livewire component with the message '{{ $message }}'</div>
        HTML;
    }
}
```

Then I use it in my template:

```blade
<livewire:static-elem message="Welcome!" />
```

Even when I have no additional code that will run once this is rendered on the server, this is the HTML output:

```html
<div wire:snapshot="{&quot;data&quot;:{&quot;message&quot;:&quot;Welcome!&quot;},&quot;memo&quot;:{&quot;id&quot;:&quot;NhoRoT5LEXWYVsyEyeWo&quot;,&quot;name&quot;:&quot;static-elem&quot;,&quot;path&quot;:&quot;hyperlink&quot;,&quot;method&quot;:&quot;GET&quot;,&quot;children&quot;:[],&quot;scripts&quot;:[],&quot;assets&quot;:[],&quot;errors&quot;:[],&quot;locale&quot;:&quot;en&quot;},&quot;checksum&quot;:&quot;7629acb206921c6b9e72e8c0d66eecaa1c03733a82f28ee34805409852ee29bd&quot;}" wire:effects="[]" wire:id="NhoRoT5LEXWYVsyEyeWo">This is a static livewire component with the message 'Welcome!'</div>
```

I wouldn't really see this as a problem if it was only a couple of components, but because we're using this Block architecture in Filament, we often are echoing out all these components in big loops so that clients can build dynamic pages with these blocks.

So you can imagine how quickly the HTML can grow when you might have a dozen of these in the page body. There are times when the snapshots even exceed the size of the actual HTML it is on.

I understand why this exists for components that have interactivity, but I'd like to be able to turn snapshots, and these extra attributes, off when I know I don't need them.

In other words, this is what I would like to see:

```html
<div>This is a static livewire component with the message 'Welcome!'</div>
```

Is there any way to do this now? I looked at the source for rendering the components, and it looks like it will always attach these attributes.

I don't expect Livewiwe to be super smart and understand whether or not I'm using interactivity, but it would be nice to control this on a component by component basis when I know that I only want to render it on the server and then do nothing else with it after.

I would just like something like this:

```php
use Livewire\Component;
use Livewire\Attributes\Static;
 
#[Static] 
class StaticElem extends Component
{
    // ...
}
```

Then I can write all my components in Livewire, keep it all in one place, still have great testing features, and even improve performance by skipping the snapshot generation.

Am I missing anything here? Is anyone else using/considering a similar approach to this for their components?

**Comment by EdgarsJoja** [comment]
I was also looking for something similar, for the same reasons. While it is not perfect, I found out that passing values through render method, instead of public properties, does not store them in snapshot object - which is exactly what I needed for static components, to keep the HTML size down.
Later found it also mentioned in the docs:
https://livewire.laravel.com/docs/components#sharing-additional-data-with-the-view

**Reply by james2doyle** [reply]
Great link! I'm not sure if it would remove the extra HTML attributes that are added but it might achieve everything else I want
